Documentation/design-docs/eh-writethru.md

```diff
@@ -0,0 +1,378 @@
+# Exception Handling Write Through Optimization.
+
+Write through is an optimization done on local variables that live across exception handling flow like a handler, filter, or finally so that they can be enregistered - treated as a register candidate - throughout a method.  For each variable live across one of these constructs, the minimum requirement is that a store to the variables location on the stack is placed between a reaching definition and any point of control flow leading to the handler, as well as a load between any return from a filter or finally and an upward exposed use.  Conceptually this maintains the value of the variable on the stack across the exceptional flow which would kill any live registers.  This transformation splits a local variable into multiple enregisterable compiler temporaries backed by the local variable on the stack. For local vars that additionally have appearances within a eh construct, a load from the stack local is inserted to a temp that will be enregistered within the handler.
+
```

Comment:
Do not underestimate the cost of stores.  They are one of the most expensive operations.  While they do not block a single thread they are definitely not free.

It would be nice to keep the stack store cost of this to exceptional cases.  Unwind would normally take core of this for us automatically, but we are using a shared frame between the Function and its exception handler.  

We should be able to do these stores only when an exception occurs.  The cost would be additional metadata structures to keep track of which register slots need to be pushed to their corresponding stack slot on an exception based on PC.